File: ProjectAllocationManagerMCP/Tools/ListEngineersTool.cs
using ModelContextProtocol.Server;
using ProjectAllocationManagerMCP.Models;
using ProjectAllocationManagerMCP.Services;
using System.ComponentModel;

namespace ProjectAllocationManagerMCP.Tools
{
    [McpServerToolType]
    public class ListEngineersTool
    {
        [McpServerTool, Description("List all engineers in the system")]
        public static async Task<List<Engineer>> ListEngineers(AllocationService allocationService)
        {
            return await allocationService.GetEngineersAsync();
        }
    }
}

File: ProjectAllocationManagerMCP/src/main/kotlin/io/modelcontextprotocol/allocation/tools/ListEngineersTool.kt
package io.modelcontextprotocol.allocation.tools

import io.modelcontextprotocol.allocation.services.AllocationService
import io.modelcontextprotocol.kotlin.sdk.CallToolRequest
import io.modelcontextprotocol.kotlin.sdk.CallToolResult
import io.modelcontextprotocol.kotlin.sdk.TextContent
import io.modelcontextprotocol.kotlin.sdk.Tool
import kotlinx.serialization.json.Json

class ListEngineersTool(
    private val allocationService: AllocationService,
    private val json: Json,
) : McpTool {
    override fun getToolDefinition() =
        ToolDefinition(
            name = "list_engineers",
            description = "List all engineers in the system with their details including ID, name, role, and skills.",
            inputSchema = Tool.Input(),
        )

    override suspend fun execute(request: CallToolRequest): CallToolResult {
        val engineers = allocationService.getEngineersAsync()
        val engineersJson = json.encodeToString(engineers)
        return CallToolResult(content = listOf(TextContent(engineersJson)))
    }
}
